lib/src/io_client.dart

```diff
@@ -23,42 +24,82 @@ class IOClient extends BaseClient {
 
   /// Sends an HTTP request and asynchronously returns the response.
   @override
-  Future<IOStreamedResponse> send(BaseRequest request) async {
+  Future<IOStreamedResponse> send(BaseRequest request, {Duration? timeout}) {
+    final completer = Completer<IOStreamedResponse>();
+    _send(request, timeout, completer);
+    return completer.future;
+  }
+
+  Future<void> _send(BaseRequest request, Duration? timeout,
+      Completer<IOStreamedResponse> completer) async {
     var stream = request.finalize();
 
+    Timer? timer;
+    late void Function() onTimeout;
+    if (timeout != null) {
+      timer = Timer(timeout, () {
+        onTimeout();
+      });
+      onTimeout = () {
+        completer.completeError(TimeoutException('Request aborted', timeout));
+      };
+    }
     try {
       var ioRequest = (await _inner!.openUrl(request.method, request.url))
         ..followRedirects = request.followRedirects
         ..maxRedirects = request.maxRedirects
         ..contentLength = (request.contentLength ?? -1)
         ..persistentConnection = request.persistentConnection;
+      if (completer.isCompleted) return;
       request.headers.forEach((name, value) {
         ioRequest.headers.set(name, value);
       });
 
+      if (timeout != null) {
+        onTimeout = () {
+          ioRequest.abort();
+          completer.completeError(TimeoutException('Request aborted', timeout));
+        };
+      }
+
       var response = await stream.pipe(ioRequest) as HttpClientResponse;
+      if (completer.isCompleted) return;
 
       var headers = <String, String>{};
       response.headers.forEach((key, values) {
         headers[key] = values.join(',');
       });
+      var responseStream = response.handleError((error) {
+        final httpException = error as HttpException;
+        throw ClientException(httpException.message, httpException.uri);
+      }, test: (error) => error is HttpException).transform<List<int>>(
+          StreamTransformer.fromHandlers(handleDone: (sink) {
+        timer?.cancel();
+        sink.close();
+      }));
+
+      if (timeout != null) {
+        onTimeout = () {
+          // TODO, is this necessary? How will it surface?
+          response.detachSocket().then((socket) => socket.destroy());
```

Comment:
Should you throw a `TimeoutException` here?

Reply:
Yes, looks like I need to handle this on the response stream.